### PR TITLE
Implement Authentication

### LINKED ISSUE #7: User Sign-in with Username/Password
- [ ] The API needs endpoints for users to login with a username and password.
- [ ] The existing Users implementation should be altered to hash passwords.
- [ ] These accounts should also be used to authenticate the endpoints.

### PR DESCRIPTION
**Pull request by ryanlyrl**
Closes #7 

## Descriptions
* Implements email/password auth, with an access token/refresh token flow.
* Also adds auto-generated Swagger-based API docs

## Limitations
Currently, the skeleton for password reset is there, but not implemented until emailing is setup. That part of the code has not been tested yet.

Ignore the massive number of commits... messed up my merge

### CODE REVIEW

File: src/controllers/authController.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+// Handles a password reset request. Responsible for taking their email and sending the reset email - see /auth/reset
+export const resetRequest = async (req: Request, res: Response): Promise<void> => {
+    const userRepository = getManager().getRepository(User);
+    const user = await userRepository.findOne({email: req.body.email});
+
+    if(user) {
+        const token = await auth.generateToken(user.uuid, 'reset');
+        // TODO: Send email here
```
Comment (dellisd) [inline src/controllers/authController.ts:48]: We should open an issue to track this.

File: src/controllers/userController.ts [hunk 2]
```diff
@@ -1,25 +1,37 @@
...
+import {validate} from 'class-validator';
+import { v4 as uuidv4 } from 'uuid';
+import * as bcrypt from 'bcrypt';
 
 export const getUsers = async (req: Request, res: Response): Promise<void> => {
     const userRepository = getManager().getRepository(User);
-    const users = await userRepository.find();
+    const users = await userRepository.find({select: ['uuid', 'email', 'firstName', 'lastName', 'role', 'githubId', 'discordId']});
     res.status(200).send(users);
 }
```
Comment (dellisd) [inline src/controllers/userController.ts:12]: Presumably this function will be put behind authentication, or removed altogether in the future, correct?

File: src/controllers/userController.ts [hunk 3]
```diff
@@ -1,25 +1,37 @@
...
+    userData.password = password;
+    userData.uuid = uuidv4();
+
+    //eslint-disable-next-line @typescript-eslint/ban-types
+    const newUser = userRepository.create({...userData} as Object); // This makes TypeORM not return an array...
     const errors = await validate(newUser);
     if(errors.length > 0) {
         res.sendStatus(400);
     } else {
         // TODO: Better validation for email uniqueness
```
Comment (dellisd) [inline src/controllers/userController.ts:29]: We can also extract an issue from this

File: src/middleware/authMiddleware.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
+import { encode, decode, TAlgorithm } from 'jwt-simple';
+import {Request, Response, NextFunction} from 'express';
+import {AccessToken} from '../entity/AccessToken';
+import {RefreshToken} from '../entity/RefreshToken';
+import {PasswordReset} from '../entity/PasswordReset';
+import {getManager} from 'typeorm';
+
+const SECRET_KEY = process.env.JWT_KEY || ''; // This must be defined, or server.ts will kill the application
```
Comment (dellisd) [inline src/middleware/authMiddleware.ts]:
```suggestion
const SECRET_KEY = process.env.JWT_KEY ?? ''; // This must be defined, or server.ts will kill the application 
```

File: src/middleware/authMiddleware.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+/**
+ * Verifies if a token a valid or not
+ * 
+ * @param token the token
+ * @param type the expected type of the token
+ * 
+ * @returns VerifyResult object, with result of the decoding and the uid of the token owner if the token was valid
+ */
+export const verifyToken = async (token: string, type: string): Promise<VerifyResult> => {
```
Comment (dellisd) [inline src/middleware/authMiddleware.ts]: Would it make sense to extract a Type to represent the `type` parameter?
e.g. something like 
```ts
type TokenType = "access" | "refresh" | "reset";
```

File: src/middleware/authMiddleware.ts [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+                uid: undefined
+            }
+        }
+    } else {
+        // Should never happen, but if we get an unrecognized type, just make it invalid
+        return {
+            result: AuthTokenStatus.Invalid, 
+            uid: undefined
+        }
+    }
```
Comment (dellisd) [inline src/middleware/authMiddleware.ts]: With a union type, this won't be necessary.

File: src/middleware/authMiddleware.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+    return {
+        result: AuthTokenStatus.Valid,
+        uid: session.session?.id
+    }
+}
+
+// Middleware for verifying if a request is authenticated via a Bearer token in Authentication header
+// TODO: Implement authorization based on role
```
Comment (dellisd) [inline src/middleware/authMiddleware.ts]: Open an issue to track this too!

File: tsconfig.json [hunk 8]
```diff
@@ -4,7 +4,7 @@
 
     /* Basic Options */
     // "incremental": true,                   /* Enable incremental compilation */
-    "target": "es2020",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
+    "target": "es6",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
```
Comment (dellisd) [inline tsconfig.json]: Any particular reason for this change?

**Review by dellisd — CHANGES_REQUESTED** [review]

File: src/controllers/userController.ts (hunk 2 above)
Comment (ryanlyrl) [reply src/controllers/userController.ts:12]: I just addressed this by adding in the authorization part. This endpoint is limited to users with the Organizer role now:
`router.get('/', auth.authenticate(Role.Organizer), user.getUsers);` 

If we end up just never using it, I'll remove this endpoint.

File: tsconfig.json (hunk 8 above)
Comment (ryanlyrl) [reply tsconfig.json]: I thought I had a bug with optional chaining, but it turns out my WSL node was just outdated for some reason (actually, for some reason the entire thing was targeting Node 12, I've upped that to Node 14). It's back to ES2020.

File: src/controllers/userController.ts (hunk 2 above)
Comment (dellisd) [reply src/controllers/userController.ts:12]: Is there test coverage for that?

Comment (ryanlyrl) [reply src/controllers/userController.ts:12]: Yup, added and passing.

**Review by dellisd — APPROVED** [review]

**Pull request merged**